module/connection/ServerDialog.kt

```diff
@@ -78,48 +91,80 @@ object ServerDialog {
 
         override fun isValid(): Boolean {
             return when (server) {
-                TYPEDB -> !address.isBlank()
-                TYPEDB_CLUSTER -> !(address.isBlank() || username.isBlank() || password.isBlank())
+                TYPEDB -> coreAddress.isNotBlank()
+                TYPEDB_CLUSTER -> !(clusterAddresses.isEmpty() || username.isBlank() || password.isBlank())
             }
         }
 
         override fun trySubmit() {
             when (server) {
-                TYPEDB -> StudioState.client.tryConnectToTypeDBAsync(address) { StudioState.client.connectServerDialog.close() }
-                TYPEDB_CLUSTER -> when {
-                    caCertificate.isBlank() -> StudioState.client.tryConnectToTypeDBClusterAsync(
-                        address, username, password, tlsEnabled
-                    ) { StudioState.client.connectServerDialog.close() }
-                    else -> StudioState.client.tryConnectToTypeDBClusterAsync(
-                        address, username, password, caCertificate
-                    ) { StudioState.client.connectServerDialog.close() }
+                TYPEDB ->  {
+                    StudioState.client.tryConnectToTypeDBAsync(coreAddress) {
+                        StudioState.client.connectServerDialog.close()
+                    }
+                }
+                TYPEDB_CLUSTER -> {
+                    when {
+                        caCertificate.isBlank() -> StudioState.client.tryConnectToTypeDBClusterAsync(
+                            clusterAddresses.toSet(), username, password, tlsEnabled
+                        ) {
+                            StudioState.client.connectServerDialog.close()
+                        }
+                        else -> StudioState.client.tryConnectToTypeDBClusterAsync(
+                            clusterAddresses.toSet(), username, password, caCertificate
+                        ) {
+                            StudioState.client.connectServerDialog.close()
+                        }
+                    }
                 }
             }
             appData.server = server
-            appData.address = address
+            appData.coreAddress = coreAddress
+            appData.clusterAddresses = clusterAddresses
             appData.username = username
             appData.tlsEnabled = tlsEnabled
             appData.caCertificate = caCertificate
         }
     }
 
+    private object CreateAddressForm : Form.State {
+
+        var value: String by mutableStateOf("")
+
+        override fun cancel() {
+            StudioState.client.manageAddressesDialog.close()
+        }
+
+        override fun isValid(): Boolean {
+            return value.isNotBlank()
+        }
+
+        override fun trySubmit() {
+            assert(isValid())
+            state.clusterAddresses.add(value)
+            value = ""
+        }
```

Comment:
Having interrogated this more, we've decided that we should refactor `Form.State` to be an abstract class, not an interface, and that `trySubmit` and `trySubmitIfValid` should simply be `submit` and `submitIfValid` respectively.

The reasoning for this is that the logic for using the term `try` when all we do is assert that `isValid()` is true for the form isn't sufficient for this function to be named `try`. When you submit a form, it may do any number of things, but there's anything we're trying before submitting apart from calling assertions. Additionally, these assertions are erased when we create release versions of the software and only exist when adding a JVM flag that we use during development.

Reply:
Done.